Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+### Imports location
+
+Always put imports at the top of the file instead of nested scopes. In other words, imports must always be grouped together in one place.
+
+## Braces
+
+### Definitions
+
+Avoid using braces for simple expressions; write this:
+
+```scala
+def square(x: Int) = x * x
+```
+
+but not this:
+
+```scala
+def square(x: Int) = {
+  x * x
+}
+```
+
+even though it may be tempting to distinguish the method body syntactically. The first alternative has less clutter and is easier to read. Avoid syntactical ceremony unless it clarifies.
+
+### Scoping and blocks
+
+It is acceptable to use braces in order to limit the namespace pollution, but this should almost always be used only when the result of the block is assigned to another variable:
+
+```scala
+val someValue = {
+  val localVariable = ...
+  ...
+}
+```
+
+Moreover, using blocks to put complex expressions in other expressions, most notable function calls, hinders readability in a really bad way and should be avoided:
+
+```scala
+// bad
+someFunction(arg1, arg2, {
+  ...more code...
+}, arg4)
+
+// good
+val arg3 = {
+  ...more code...
+}
+someFunction(arg1, arg2, arg3, arg4)
+```
+
+## Comments
+
+Use Scaladoc to provide API documentation. Use the following style:
+
+```scala
+/**
+ * ServiceBuilder builds services 
+ * ...
+ */
+```
+
+but not the standard Scaladoc style:
+
+```scala
+/** ServiceBuilder builds services
+  * ...
+  */
+```
+
+Public API must always be documented with Scaladoc.
+
+Write comments to clarify parts of logic which are not obvious, but *only* when it is absolutely impossible to rewrite the logic to be obvious in the first place. In general, try follow the rule "readable code comments itself".
+
+`TODO:`-like comments are acceptable when they are really necessary but should not be overused.
+
+# Type system
+
+The primary objective of a powerful static type system is to reduce the number of bugs and other programming errors. Moreover, static type system often gives benefits in code readability and navigation (with the help of IDEs). However, overusing the type system in Scala leads to code which is very hard to understand. Therefore, most of the advanced techniques which are possible with the Scala type system must be avoided.
+
+## Return type annotations
+
+Scala allows one to omit return type annotations on `def`s. However, these annotations increase readability and help to avoid many issues with the type system, therefore all public-facing methods must have return types:
+
+```scala
+trait PublicInterface {
+  def computeValue(input: String): AnotherEntity
+}
+```
+
+This includes overrides in the classes extending the supertype:
+
+```scala
+class PublicInterfaceImpl extends PublicInterface {
+  override def computeValue(input: String): AnotherEntity = {
+    ...
+  }
+}
+```
+
+It is acceptable to omit return types for *short* local functions, but only when it is very clear what the return type is, for example:
```

Comment:
I'd suggest to allow omitting return types for short private functions as well, because they're never part of any interface and usually serve for similar purposes as well as local functions.